File: server/controllers/historyController.js
const { HistoryRepository } = require('../models/dbStore');

// @desc    Get user preparation history
// @route   GET /api/history
const getHistory = async (req, res, next) => {
  try {
    const cursor = await HistoryRepository.find({ userId: req.user._id });
    const sorted = cursor.sort ? cursor.sort({ createdAt: -1 }) : cursor;
    const history = sorted.limit ? await sorted.limit(50) : sorted;

    res.status(200).json({
      success: true,
      count: history.length,
      history,
    });
  } catch (error) {
    next(error);
  }
};

// @desc    Delete single history entry
// @route   DELETE /api/history/:id
const deleteHistoryItem = async (req, res, next) => {
  try {
    const item = await HistoryRepository.findOneAndDelete({
      _id: req.params.id,
      userId: req.user._id,
    });

    if (!item) {
      return res.status(404).json({
        success: false,
        message: 'History item not found.',
      });
    }

    res.status(200).json({
      success: true,
      message: 'History item removed.',
    });
  } catch (error) {
    next(error);
  }
};

// @desc    Clear all user history
// @route   DELETE /api/history
const clearHistory = async (req, res, next) => {
  try {
    await HistoryRepository.deleteMany({ userId: req.user._id });

    res.status(200).json({
      success: true,
      message: 'All preparation history cleared.',
    });
  } catch (error) {
    next(error);
  }
};

module.exports = {
  getHistory,
  deleteHistoryItem,
  clearHistory,
};

File: server/src/main/java/com/nexoffer/controller/HistoryController.java
package com.nexoffer.controller;

import com.nexoffer.model.History;
import com.nexoffer.model.User;
import com.nexoffer.repository.HistoryRepository;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/history")
public class HistoryController {

    private final HistoryRepository historyRepository;

    public HistoryController(HistoryRepository historyRepository) {
        this.historyRepository = historyRepository;
    }

    @GetMapping
    public ResponseEntity<?> getHistory(@AuthenticationPrincipal User user) {
        List<History> list = historyRepository.findTop50ByUserIdOrderByCreatedAtDesc(user.getId());
        return ResponseEntity.ok(Map.of(
                "success", true,
                "count", list.size(),
                "history", list
        ));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteHistoryItem(@AuthenticationPrincipal User user, @PathVariable Long id) {
        Optional<History> itemOpt = historyRepository.findByIdAndUserId(id, user.getId());
        if (itemOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("success", false, "message", "History item not found."));
        }

        historyRepository.delete(itemOpt.get());
        return ResponseEntity.ok(Map.of("success", true, "message", "History item removed."));
    }

    @DeleteMapping
    public ResponseEntity<?> clearHistory(@AuthenticationPrincipal User user) {
        historyRepository.deleteByUserId(user.getId());
        return ResponseEntity.ok(Map.of("success", true, "message", "All preparation history cleared."));
    }
}
